src/sinks/gcp/cloud_storage.rs

```diff
@@ -0,0 +1,543 @@
+        let bucket = config.bucket.clone();
+        Ok(GcsSink {
+            client,
+            creds,
+            settings,
+            base_url,
+            bucket,
+        })
+    }
+
+    fn service(self, config: &GcsSinkConfig, cx: &SinkContext) -> crate::Result<RouterSink> {
+        let request = config.request.unwrap_with(&REQUEST_DEFAULTS);
+        let encoding = config.encoding.clone();
+
+        let compression = match config.compression {
+            Compression::Gzip => true,
+            Compression::None => false,
+        };
+        let batch = config.batch.unwrap_or(bytesize::mib(10u64), 300);
+
+        let key_prefix = if let Some(kp) = &config.key_prefix {
+            Template::from(kp.as_str())
+        } else {
+            Template::from("date=%F/")
+        };
+
+        let settings = self.settings.clone();
+
+        let svc = ServiceBuilder::new()
+            .map(move |req| RequestWrapper::new(req, settings.clone()))
+            .settings(request, GcsRetryLogic)
+            .service(self);
+
+        let sink = crate::sinks::util::BatchServiceSink::new(svc, cx.acker())
+            .partitioned_batched_with_min(PartitionBuffer::new(Buffer::new(compression)), &batch)
+            .with_flat_map(move |e| iter_ok(encode_event(e, &key_prefix, &encoding)));
+
+        Ok(Box::new(sink))
+    }
+
+    pub fn healthcheck(&self) -> crate::Result<Healthcheck> {
+        let mut builder = Request::builder();
+        builder.method(Method::HEAD);
+        builder.uri(self.base_url.parse::<Uri>()?);
+
+        let mut request = builder.body(Body::empty()).unwrap();
+        if let Some(creds) = &self.creds {
+            creds.apply(&mut request);
+        }
+
+        let healthcheck =
+            self.client
+                .request(request)
+                .map_err(Into::into)
+                .and_then(healthcheck_response(
+                    self.creds.clone(),
+                    GcsError::BucketNotFound {
+                        bucket: self.bucket.clone(),
+                    }
+                    .into(),
+                ));
+
+        Ok(Box::new(healthcheck))
+    }
+}
+
+impl Service<RequestWrapper> for GcsSink {
+    type Response = hyper::Response<Body>;
+    type Error = hyper::Error;
+    type Future = hyper::client::ResponseFuture;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        Ok(().into())
+    }
+
+    fn call(&mut self, request: RequestWrapper) -> Self::Future {
+        let settings = request.settings;
+
+        let uri = format!("{}{}", self.base_url, request.key)
+            .parse::<Uri>()
+            .unwrap();
+        let mut builder = Request::builder();
+        builder.method(Method::PUT);
+        builder.uri(uri);
+        let headers = builder.headers_mut().unwrap();
+        headers.insert("content-type", HeaderValue::from_str("text/plain").unwrap()); // FIXME needs proper content type
```

Comment:
Is this fixme still valid?

Reply:
Absolutely. I missed applying the content type in the settings to the request. Will fix.